entropy-mnemonic conversions

sobuch opened the pull request.

sobuch committed:
First version entropy-mnemonic conversions

sobuch commented:
Should we include test cases for private functions? (its kind of design antipatern) If so, how to do it with current includes and so on?

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -67,8 +92,6 @@ def __is_valid_entropy(entropy: bytes) -> bool:
 
 def __is_valid_mnemonic(mnemonic: str) -> bool:
     """Check whether provided string represents a valid mnemonic phrase based on current dictionary.
-    Currently uses 1 default dictionary with English words.
-    # TODO Should we support multiple dictionaries for various languages?
```
Why were these lines removed when the *TODO* was not solved?

sobuch replied:
I moved TODO to more appropriate location, other functions only call function to get dictionary, proposed changes wont be in those other methods

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+
+    """
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'english.txt'), 'r') as f:
+        l = f.read().splitlines()
```
What errors/exceptions can happen during opening and reading a file? Will this function catch them or will they get propagated? Consider adding docstring `:raises:`.

Please consider errors we discuseed, . Who should handle them? If function's caller, then please document exceptions or catch and reraise our own subclassed.
- file not found
- no permissions for file
- dictionary does not contain exactly 2048 lines
- dictionary is too big (2048 lines OK, but too long words) like hundreds of MB...
- every line has exactly 1 word (no whitespaces)

sobuch replied:
jop this is why its WIP :) , will get back to this

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+    shift = len(entropy) // 4
+    entropy_bin = (int.from_bytes(entropy, byteorder='big') << shift) |\
+                  (int.from_bytes(checksum, byteorder='big') >> 256 - shift)
+    indexes = [(entropy_bin >> i * 11) & 2047 for i in reversed(range(shift * 3))]
```
I find this part hard to read. If I understand it correctly, `checksum` is not checksum yet, it's just hash of entropy, while `int.from_bytes(checksum, byteorder='big') >> 256 - shift` is checksum. This way, `entropy_bin` is entropy with appended checksum. I see that last line `[(entropy_bin >> i * 11) & 2047 for i in reversed(range(shift * 3))]` should divide `entropy_bin` into parts of 11 bits, but what is `shift * 3`? It would be really helpful if you could reference [specification](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic) and add comments.

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
 
 def __mnemonic2entropy(mnemonic: str) -> bytes:
     """Convert mnemonic phrase to entropy using dictionary.
-    Currently uses 1 default dictionary with English words.
-    # TODO Should we support multiple dictionaries for various languages?
```
Why were these lines removed when the *TODO* was not solved?

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+    l = (len(words))
+    indexes = [word_dict[word] for word in words]
+    entropy_bin = sum([indexes[-i - 1] << i * 11 for i in reversed(range(l))]) << l - (l // 8) * 8
+    return entropy_bin.to_bytes((l * 11 + 7) // 8, byteorder='big')
```
Could you please add some comments explaining operations on these lines?

lsolodkova commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+    :return: List and dictionary of words
+
+    """
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'english.txt'), 'r') as f:
```
Some checks should be here: what if file wasn't opened, what if it hasn't a structure we need, what if we've read more or less than we need, etc.

mvondracek replied:
Same as here, it should raise exception, this function cannot recover from such error, the caller should handle it. https://github.com/mvondracek/PA193_mnemonic_Slytherin/pull/19#discussion_r334272962

mvondracek commented:
> Should we include test cases for private functions? (its kind of design antipatern) If so, how to do it with current includes and so on?

If there are tests for private functions, they depend on current implementation. If you change the private implementation, these tests will break. For example tests for `_secure_seed_compare` are more like doctests and can definitely be transformed into proper doctests if we do not want to test private functions. There is also question, whether everything except `generate`, `recover`, and `verify` should be private...

If you want to test private function, single underscore allows simple imports.

lsolodkova commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
     """
-    pass
+    word_list, _ = __get_dictionary()
+    checksum = hashlib.sha256(entropy).digest()
```
Need to check that entropy has the correct length.

sobuch replied:
I believe it should be checked before its passed to this function

lsolodkova commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
     """
-    pass
+    _, word_dict = __get_dictionary()
+    words = mnemonic.split()
```
need to check the mnemonic for appropriate length. And what if it contains some word out of current dictionary?

sobuch replied:
same as https://github.com/mvondracek/PA193_mnemonic_Slytherin/pull/19#discussion_r334274813

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,47 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
-    # TODO Should we support multiple dictionaries for various languages?
     :rtype: bytes
-    :return: Entropy
+    :return: Entropy with appended checksum in the last byte
```
Should it return entropy **including** checksum? If so, `__entropy2mnemonic` and `__mnemonic2entropy` are not symmetric.

sobuch replied:
Depends on whether we want to check the checksum inside the function or outside. The way it is implemented now, function should never fail and we dont have to consider anything related to that.

mvondracek commented:
Please see  #22, thanks.

sobuch committed:
corrections for e2m and m2e

sobuch committed:
remove extra calls to is_valid_mnemonic

sobuch committed:
check dictionary

lsolodkova commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,97 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+    l = len(words)
+    
+    try:
+        indexes = [word_dict[word] for word in words]
```
As far as I understand, the check of mnemonics is implemented right inside the function, here and below by calculating the checksum. But what about the allowed length of the mnemonic sentence and, so, the entropy?

sobuch replied:
should be also checked now :)

mvondracek replied:
Would be solved by #26. But that would be for a separate PR, don't worry. 😄

sobuch committed:
add checks for mnemonic

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -132,8 +201,6 @@ def verify(mnemonic: str, expected_seed: bytes, seed_password: str = '') -> bool
     :rtype: bool
     :return: True if provided phrase generates expected seed, False otherwise.
     """
-    if not __is_valid_mnemonic(mnemonic):
```
Validation of mnemonic is removed here, and unchecked `mnemonic` is passed to `__generate_seed` which does not perform any validation at the moment. This approach could be fine, but this PR might sync with *dev* branch when #21 is merged so that `__generate_seed` actually does any checks.

However, as of ebaa9ed6a277b522167eda3c7be2e60f1644e0ed,  `__generate_seed` **does not validate** mnemonic. 

https://github.com/mvondracek/PA193_mnemonic_Slytherin/blob/04ab60aab39a998ace40d9c73cabaee988474e65/PA193_mnemonic_Slytherin/mnemonic.py#L24-L35

mvondracek replied:
Would be solved by #26.

mvondracek replied:
But that would be for a separate PR, don't worry. 😄

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -38,24 +40,102 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+
+    """
+    l = []
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'english.txt'), 'r') as f:
```
Please place filename to parameter with default value so we could write tests for various *malicious* dictionaries.
~~~py
ENGLISH_DICTIONARY_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'english.txt')


def __get_dictionary(file_path: str=ENGLISH_DICTIONARY_PATH) -> Tuple[List[str], Dict[str, int]]:
~~~

mvondracek commented on PA193_mnemonic_Slytherin/mnemonic.py:
```diff
@@ -116,9 +188,6 @@ def recover(mnemonic: str, seed_password: str = '') -> Tuple[bytes, bytes]:
     :return: Two item tuple where first is initial entropy and second is seed.
     """
-    if not __is_valid_mnemonic(mnemonic):
-        raise ValueError('invalid mnemonic')
```
Would be solved by #26.

mvondracek replied:
But that would be for a separate PR, don't worry. 😄

mvondracek reviewed (changes requested):
Related #26.

sobuch committed:
parametrize get_dictionary

mvondracek commented:
Waits until #21 is merged. https://github.com/mvondracek/PA193_mnemonic_Slytherin/pull/19#discussion_r335148843

mvondracek committed:
Merge branch 'dev' into task-11-entropy-mnemonic-transformations

mvondracek committed:
remove `is_valid_mnemonic` from mnemoniccli

Function `__is_valid_mnemonic`/`_is_valid_mnemonic`/`is_valid_mnemonic`
was removed in 545d16133398337a5e3cf9f2315f23276f4c7c7d, then
0d2b5559775a08cae203868a315752e6ebb10eed removed calls to this function
inside `mnemonic` module.
This commit removes calls from mnemoniccli and reacts to raised
exceptions.

mvondracek committed:
remove `is_valid_seed` from mnemoniccli, EAFP

mvondracek committed:
remove `is_valid_entropy` from mnemoniccli, EAFP

mvondracek committed:
cleanup

mvondracek committed:
fix: removed `is_valid_mnemonic` from package init

related 0d2b5559775a08cae203868a315752e6ebb10eed

mvondracek commented:
Updated from dev. Updated the program (`mnemoniccli`) and package init, because `is_valid_mnemonic` was removed. The program now uses EAFP instead of LBYL. Tests are [OK now](https://travis-ci.org/mvondracek/PA193_mnemonic_Slytherin/builds/600106016).

mvondracek committed:
add `test_verify_invalid_mnemonic_too_long`, failing until bug in `verify` is fixed

Too long mnemonic which was not validated causes `OverflowError:
password is too long.` in `pbkdf2_hmac` during `_generate_seed`.

mvondracek commented:
Detected bug #28. Added test in f0b267c, which will be failing until the bug is fixed.
~~~
ERROR: test_verify_invalid_mnemonic_too_long (PA193_mnemonic_Slytherin.test_mnemonic.TestMnemonicPublic)

Too long mnemonic phrase which should not be propagated to BPKDF2.

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/mvondracek/PA193_mnemonic_Slytherin/PA193_mnemonic_Slytherin/test_mnemonic.py", line 217, in test_verify_invalid_mnemonic_too_long

    verify(mnemonic, seed)

  File "/home/travis/build/mvondracek/PA193_mnemonic_Slytherin/PA193_mnemonic_Slytherin/mnemonic.py", line 217, in verify

    generated_seed = _generate_seed(mnemonic, seed_password)

  File "/home/travis/build/mvondracek/PA193_mnemonic_Slytherin/PA193_mnemonic_Slytherin/mnemonic.py", line 39, in _generate_seed

    return pbkdf2_hmac('sha512', mnemonic, passphrase, PBKDF2_ROUNDS, SEED_LEN)

OverflowError: password is too long.
~~~

sobuch commented:
superceded by https://github.com/mvondracek/PA193_mnemonic_Slytherin/pull/36
